impeller/entity/contents/color_source_contents.h

```diff
@@ -100,6 +101,95 @@ class ColorSourceContents : public Contents {
+                    RenderPass& pass,
+                    const PipelineBuilderMethod& pipeline_builder,
+                    typename VertexShaderT::FrameInfo frame_info,
+                    const BindFragmentCallback& bind_pipeline_callback) const {
+    auto options = OptionsFromPassAndEntity(pass, entity);
+
+    // If overdraw prevention is enabled (like when drawing stroke paths), we
+    // increment the stencil buffer as we draw, preventing overlapping fragments
+    // from drawing. Afterwards, we need to append another draw call to clean up
+    // the stencil buffer (happens below in this method).
+    if (geometry_result.prevent_overdraw) {
+      options.stencil_mode =
+          ContentContextOptions::StencilMode::kLegacyClipIncrement;
+    }
+    options.primitive_type = geometry_result.type;
+    pass.SetVertexBuffer(std::move(geometry_result.vertex_buffer));
+    pass.SetStencilReference(entity.GetClipDepth());
+
+    frame_info.depth = entity.GetShaderClipDepth();
+    frame_info.mvp = geometry_result.transform;
+
+    VertexShaderT::BindFrameInfo(
+        pass, renderer.GetTransientsBuffer().EmplaceUniform(frame_info));
+
+    std::shared_ptr<Pipeline<PipelineDescriptor>> pipeline =
+        (renderer.*pipeline_builder)(options);
+    pass.SetPipeline(pipeline);
+
+    if (!bind_pipeline_callback(pass)) {
+      return false;
+    }
+
+    if (!pass.Draw().ok()) {
+      return false;
+    }
+
+    // If we performed overdraw prevention, a subsection of the clip heightmap
+    // was incremented by 1 in order to self-clip. So simply append a clip
+    // restore to clean it up.
+    if (geometry_result.prevent_overdraw) {
+      auto restore = ClipRestoreContents();
+      restore.SetRestoreCoverage(GetCoverage(entity));
+      return restore.Render(renderer, entity, pass);
+    }
+    return true;
+  }
+
+  template <typename VertexShaderT>
+  bool DrawPositions(const ContentContext& renderer,
+                     const Entity& entity,
+                     RenderPass& pass,
+                     const PipelineBuilderMethod& pipeline_builder,
+                     typename VertexShaderT::FrameInfo frame_info,
+                     const BindFragmentCallback& bind_pipeline_callback) const {
+    auto geometry_result =
+        GetGeometry()->GetPositionBuffer(renderer, entity, pass);
+
+    return DrawGeometry<VertexShaderT>(geometry_result, renderer, entity, pass,
+                                       pipeline_builder, frame_info,
+                                       bind_pipeline_callback);
+  }
+
+  template <typename VertexShaderT>
+  bool DrawPositionsAndUVs(
+      Rect texture_coverage,
+      Matrix effect_transform,
+      const ContentContext& renderer,
+      const Entity& entity,
+      RenderPass& pass,
+      const PipelineBuilderMethod& pipeline_builder,
+      typename VertexShaderT::FrameInfo frame_info,
+      const BindFragmentCallback& bind_pipeline_callback) const {
```

Comment:
Alternative suggestion, make color source contents have a virtual `BindFragmentCallback` method and make this call it. If we have to pass a closure for non trivial cases then its going to end up closing over the entire contents anyway.

Reply:
Seems like a good approach. Done.

Reply:
Uhh actually applying this to my follow-up patch is a painful because we have multiple rendering branches for all of the gradients that need different binding behavior.

In all of the use cases we can capture the things we need by reference, and my understanding is that assigning a lambda to a stack `std::function` shouldn't heap allocate as long as the set of captures is reasonably small.

One sec, gonna check this...

Reply:
The different gradients are basically different content classes though
if the contents is less verbose with this change it would make more sense to split them in two IMO.

Reply:
Briefly discussed offline, we can just switch this out later if necessary. Some legit concerns over expanding our use of `std::function`, like annoying stack traces, compiler differences, etc.